### ISSUE TITLE
Solver + Driver Co-location Epic

### ISSUE THREAD

**Issue by nlordell** [cowprotocol/services#830]
This issue is a refresh of https://github.com/cowprotocol/services/issues/337, which lays out most of the remaining steps for Solver + Driver co-location (the original design was specified in https://github.com/cowprotocol/services/issues/230).

### Moving Driver Run-Loop Into the Autopilot

This section captures the work of moving the `solver::driver` run-loop into the `autopilot`, as well as adapt it to the new commit-reveal scheme. In order to do this with the least amount of down-time possible:

- [x] #1983
- [x] #855
    **Discussion point**: This allows for us to change when/where we cut new auctions in the `autopilot` without running into issues
- [x] #861
- [x] #879
- [x] #862
- [x] #863
- [x] #880
    - [x] #881
- [x] #864
- [x] #865
- [x] #877
- [x] #876 
- [ ] #885
    - [ ] #886

### Supporting Driver Interface For Existing Solvers

- [x] Run the driver + Quasimodo instance.
- [x] #831
    - **Make sure to include external solver teams in discussion** but we can start with an "initial spec" that suits our needs to unblock following tasks
- [x] #882
- [x] Run driver + solver setup in our infrastructure
- [x] #887

### Price Estimation With Drivers

- [x] #858
- [x] #872
    - [x] #873
- [x] Move all price estimators to external `driver` estimators.
- [ ] #1984

### Driver Crate

- [x] https://github.com/cowprotocol/services/issues/892

### Nice To Haves

- [x] #883
- [ ] https://github.com/cowprotocol/services/issues/545 (essentially, the autopilot should synchronize starting a run-loop when it observes a new block)
- [ ] Enable RFQ support in the driver.
    - [ ] Add new kind of "delayed interaction" in the response from HTTP solvers
- [ ] Easy to setup dev environment.
- [x] #888

**Comment by nlordell** [comment]
> I'd like to really nail down the requirements for the driver application. I think it's quite important for us to understand exactly what it's supposed to do, i.e. gather the requirements for it.

For this, let me define some more exact terms:
- _protocol node_: A component that is responsible for comunicating with the `orderbook` and `autopilot` for price estimation and settlement finding (I called it node for analogy with Ethereum nodes). It is responsible for, in a block box way, answering the following HTTP requests:
  - `/quote` for price estimation requirements in the protocol
  - `/solve` for proposing a settlement (without revealing the exact execution)
  - `/execute` for executing a proposed settlement if it wins.
- _driver_: This is, in my Ethereum node a reference implementation for the, the "execution layer" of a _protocol node_ - i.e. the component which does a lot of the shared work that all _protocol nodes_ need. This is what the `driver` binary is intended to be. Its responsibilities are:
  - Starting an HTTP server to handle the requests as expected by the `protocol` and `orderbook`.
  - Shared code for collecting existing liquidity (the main motivation here is to simplify onboarding for new solvers, also needed in some of our existing solvers).
  - Providing sanity checks for many of the responses from the solvers themselves. Note that there are protocol offenses that are punishable by slashing and removing funds from a bonding pool. The reference implementation checks these things to provide "safety-guards" for solvers. This does not replace actual post-settlement validation, but instead helps prevent existing solvers from making certain kinds of mistakes
  - Shared code for executing transactions on-chain (we have "settlement submission strategies" that do things like re-submit settlements with increasing gas costs, send cancellation transactions when they start to simulate as failure as a gas saving, etc.)
  - Providing a reference implementation for more advanced teams that want to implement their own "execution layer" in the _protocol node_.
- _searcher_: This is the component that specializes in taking an abstract "auction instance" (orders + liquidity; the optimization problem we are trying to solve) and computing an abstract settlement (clearing price map + executed orders + on-chain interactions).
  - Provides a very simple `/solve` endpoint that takes an abstract "auction instance"
  - This is the "special sauce" that current external solvers are adding.

> Possibly instead of "sell_with_fees" and "sell_without_fees" we should use a boolean flag or something, I will think a bit more about this. I think the answer will reveal itself with a little more time.

I don't think the `fees` thing really matters here (that is more important for order quoting and not price quoting). I think you only need a `side`:
```ts
{
  "sell_token": string, // Sell token address.
  "buy_token": string, // Buy token address.
  "side": "sell" | "buy",
  "amount": string, // amount to sell for "sell" orders or amount to buy for "buy" orders
}
```

> As a response, it should return the following:

I wouldn't return `gas_price` or `sell_token_price` as the protocol is sensitive to them. I would also include some "trade calldata" in order to build to be able to simulate these quotes in the `orderbook` and `protocol` binaries.

**Comment by nlordell** [comment]
> Why do we call them "interactions" instead of just "calls"? That name struck me as surprising. I haven't heard that term being used before.

This is what we call them [in the smart contract](https://github.com/cowprotocol/contracts/blob/d043b0bfac7a09463c74dfe1613d0612744ed91c/src/contracts/GPv2Settlement.sol#L125).

> IIUC the main benefit of this is privacy for the solvers, is there anything else?

One reason is definitely that we expect solvers to not want to show their hand unless they win. The other important reason is so that we build a system that is compatible with RFQ liquidity, where you first get a quote, and only get calldata if you commit to executing it. This is important for integrating with private market maker RPQ systems.

> The only thing I would add here is that I would attach an ID to the settlement summary returned from /solve. In theory /execute could simply execute the latest solution, however as a sanity check and to avoid potential timing or other unpredictable issues, I would still pass the aforementioned ID to the /execute endpoint.

💯 - I _think_ we already have this, but completely agree that we should add it if we don't already. It is also important to pass additional contextual data with the `/execute` call:
- We can "sign the win" in the protocol so that the external solver has cryptographic proof that it was asked to execute in case of a dispute
- We can include the block deadline for getting the settlement on-chain

> This endpoint should use the configured private key to broadcast the settlement transaction

I don't think we want to broadcast the transaction on `/solve` - only return "what the solver wants to do". The `protocol` then decides the winner and calls `/execute` which then broadcasts the transaction.

> and return the transaction ID to the caller

I would be careful with conflating an "execution identifer" with an Ethereum transaction hash. They are not necessarily the same:
- A transaction hash changes when a solver sends a transaction with an increased gas price to try and mine if quicker if it finds it is taking too long.
- A solver may chose to use a transaction relay service, where the transaction hash cannot be known up-front.

This is why we wanted to include `address`, `nonce`, and `calldata` for this.

> I also don't think that the solver name needs to be passed in the input, since solution IDs should be unique anyway, but passing it would be more explicit and I don't think it'd hurt either - could be useful as another sanity check. OTOH none of the other endpoints accept a solver name... ultimately my call would probably be to just leave it out and we can add it later if needed.

Are you referring to the current `driver` implementation? I think we pass in the solver name because it acts as a "mux" for `N` solvers so that they can share things like indexed liquidity. The solver name is just used for routing and an implementation detail. From the `protocol` and `orderbook`'s point of view - it is just part of the URL but has no special meaning.

> There is a small question here: should we support multiple private keys for multiple solvers

Yes, each solver should have their own private key.

> If we find that both are needed, we could support both approaches.

I think supporting multiple private keys gets this for free - we just configure multiple solvers to use the same one.

> If `/execute` is called by protocol, but doesn't publish within the deadline, does the solver get penalized? Is this the plan? How do we define the deadline?

Yes

>  Is it expressed in the number of blocks, or in time units? I suppose the deadline is basically going to be one block, since we intend to align auction cuts with blocks, right?

Blocks, always blocks 😛. I think 1 block isn't reasonable. Right now we have a 2min deadline for mining transactions, so that means 10 blocks. We would then align the following auction either:
- As soon as we observe the previous settlement
- The deadline block

>  I suppose we should continue supporting these two notifications, this should be simple enough by having a /notify endpoint which accepts the notification and the solver name, and forwards the notification to the appropriate solver.

I also think this is fine (which is why I proposed it). I don't feel strongly either way though.

**Comment by vkgnosis** [comment]
Internal deployments in our cluster are not reachable by the outside so they are safe without authentication. External driver deployments offer us a publically reachable URL that anyone on the internet could talk to. In that case it is a good idea to have authentication so there is no chance of receiving fake autopilot requests.

**Comment by MartinquaXD** [comment]
>  Enable RFQ support in the driver.
Add new kind of "delayed interaction" in the response from HTTP solvers

I'm not sure what is required here. Doesn't the fact that you can win the solver competition without providing `call_data` for it and only finalize the solution on `/reveal` already RFQ support in and of itself? Or does this point mean that the driver should also collect some RFQ liquidity before sending the request to the solver (not sure how doable this is)?
